reader_concurrency_semaphore.cc

```diff
@@ -1558,6 +1592,24 @@ void reader_concurrency_semaphore::dequeue_permit(reader_permit::impl& permit) {
     _permit_list.push_back(permit);
 }
 
+void reader_concurrency_semaphore::on_permit_timeout(reader_permit::impl& permit) {
+    try {
+        maybe_timeout_permit(_wait_list.back(), _wait_list.size());
+    } catch(const named_semaphore_timed_out& ex) {
+        while (true) {
+            auto& front = _wait_list.front();
+            if (&front == &permit) {
+                break;
+            }
+
+            front.aux_data().pr.set_exception(std::make_exception_ptr(ex));
+            dequeue_permit(front);
+        }
```

Comment:
What is this loop for?

Reply:
To remove reads from the front of the wait list when a read times out while being in the queue.

Reply:
I think this should be done in `maybe_admit_waiters()`. There we should inspect the candidate for admission and decide whether to abort/admit, then continue with next candidate until we can do neither.